File: java/src/main/java/vsdk/toolkit/galerkin/GalerkinRadianceMethod.java
/**
Galerkin radiosity, with the following variants:
- With or without hierarchical refinement
- With or without clusters
- With Jacobi, Gauss-Seidel or South well iterations
- With potential-driven or not
*/

package vsdk.toolkit.galerkin;

import java.io.OutputStream;
import java.util.ArrayList;
import java.util.Locale;
import vsdk.toolkit.common.ColorRgb;
import vsdk.toolkit.common.Error;
import vsdk.toolkit.common.RenderOptions;
import vsdk.toolkit.common.linealAlgebra.Numeric;
import vsdk.toolkit.common.linealAlgebra.Vector3D;
import vsdk.toolkit.common.statistics.Statistics;
import vsdk.toolkit.io.wrl.VrmlWriter;
import vsdk.toolkit.material.BsdfComponent;
import vsdk.toolkit.material.XxdfComponentFlag;
import vsdk.toolkit.numericalAnalysis.PatchVisitor;
import vsdk.toolkit.scene.Camera;
import vsdk.toolkit.scene.RadianceMethod;
import vsdk.toolkit.scene.RadianceMethodAlgorithm;
import vsdk.toolkit.scene.Scene;
import vsdk.toolkit.skin.Element;
import vsdk.toolkit.skin.Patch;
import vsdk.toolkit.tonemap.ToneMap;
import vsdk.toolkit.tonemap.ToneMappingContext;

public final class GalerkinRadianceMethod extends RadianceMethod {
    private static final int STRING_LENGTH = 2000;
    private static final int ALL_XXDF_COMPONENTS =
        XxdfComponentFlag.DIFFUSE_COMPONENT
            | XxdfComponentFlag.GLOSSY_COMPONENT
            | XxdfComponentFlag.SPECULAR_COMPONENT;
    private static final int ALL_BSDF_COMPONENTS =
        BsdfComponent.BRDF_DIFFUSE_COMPONENT
            | BsdfComponent.BRDF_GLOSSY_COMPONENT
            | BsdfComponent.BRDF_SPECULAR_COMPONENT
            | BsdfComponent.BTDF_DIFFUSE_COMPONENT
            | BsdfComponent.BTDF_GLOSSY_COMPONENT
            | BsdfComponent.BTDF_SPECULAR_COMPONENT;

    private Object gatheringStrategy;

    public static GalerkinState galerkinState = new GalerkinState();

    public static void freeMemory() {
        if ( GalerkinRadianceMethod.galerkinState.scratch != null ) {
            GalerkinRadianceMethod.galerkinState.scratch = null;
        }
    }

    public static void recomputePatchColor(Patch patch) {
        if ( patch == null || !(patch.radianceData instanceof GalerkinElement) ) {
            return;
        }

        GalerkinElement element = (GalerkinElement)patch.radianceData;
        ColorRgb reflectivity = element.Rd;
        ColorRgb radVis = new ColorRgb();

        // Compute the patches color based on its radiance + ambient radiance if desired
        if ( galerkinState.useAmbientRadiance != 0 ) {
            radVis.scalarProduct(reflectivity, galerkinState.ambientRadiance);
            radVis.add(radVis, element.radiance[0]);
            ToneMap.radianceToRgb(radVis, patch.color, galerkinState.toneMapOptions);
        }
        else {
            ToneMap.radianceToRgb(element.radiance[0], patch.color, galerkinState.toneMapOptions);
        }
        patch.computeVertexColors();
    }

    public GalerkinRadianceMethod() {
        className = RadianceMethodAlgorithm.GALERKIN;
        gatheringStrategy = null;
    }

    @Override
    public String getRadianceMethodName() {
        return "Galerkin";
    }

    public void setStrategy() {
        if ( galerkinState.clustered != 0 ) {
            gatheringStrategy = new Object();
        }
        else {
            gatheringStrategy = new Object();
        }
    }

    @Override
    public void parseOptions(int[] argc, String[] argv) {
        if ( argc != null || argv != null ) {
            // Nothing to parse in this migration step.
        }
    }

    private static void patchInit(Patch patch) {
        if ( patch == null || !(patch.radianceData instanceof GalerkinElement) ) {
            return;
        }

        GalerkinElement element = (GalerkinElement)patch.radianceData;
        ColorRgb reflectivity = element.Rd;
        ColorRgb selfEmittanceRadiance = element.Ed;

        if ( galerkinState.useConstantRadiance != 0 ) {
            // See Neumann et-al, "The Constant Radiosity Step", Euro-graphics Rendering Workshop
            // '95, Dublin, Ireland, June 1995, p 336-344
            element.radiance[0].scalarProduct(reflectivity, galerkinState.constantRadiance);
            element.radiance[0].add(element.radiance[0], selfEmittanceRadiance);
            if ( galerkinState.galerkinIterationMethod == GalerkinIterationMethod.SOUTH_WELL ) {
                element.unShotRadiance[0].subtract(element.radiance[0], galerkinState.constantRadiance);
            }
        }
        else {
            element.radiance[0].set(selfEmittanceRadiance.r, selfEmittanceRadiance.g, selfEmittanceRadiance.b);
            if ( galerkinState.galerkinIterationMethod == GalerkinIterationMethod.SOUTH_WELL ) {
                element.unShotRadiance[0].set(element.radiance[0].r, element.radiance[0].g, element.radiance[0].b);
            }
        }

        if ( galerkinState.importanceDriven != 0 ) {
            switch ( galerkinState.galerkinIterationMethod ) {
                case GAUSS_SEIDEL:
                case JACOBI:
                    element.potential = patch.directPotential;
                    break;
                case SOUTH_WELL:
                    element.potential = patch.directPotential;
                    element.unShotPotential = patch.directPotential;
                    break;
                default:
                    Error.fatal(-1, "patchInit", "Invalid iteration method");
            }
        }

        recomputePatchColor(patch);
    }

    @Override
    public void initialize(Scene scene, ToneMappingContext toneMapOptions) {
        galerkinState.toneMapOptions = toneMapOptions;
        if ( galerkinState.toneMapOptions == null ) {
            Error.fatal(-1, "GalerkinRadianceMethod::initialize", "Tone mapping context not provided");
        }

        galerkinState.iterationNumber = 0;
        galerkinState.cpuSeconds = 0.0f;

        GalerkinElement.initializeBasis();

        galerkinState.constantRadiance = Statistics.instance().radiance.estimatedAverageRadiance;
        if ( galerkinState.useConstantRadiance != 0 ) {
            galerkinState.ambientRadiance.clear();
        }
        else {
            galerkinState.ambientRadiance = Statistics.instance().radiance.estimatedAverageRadiance;
        }

        for ( int i = 0; scene.patchList != null && i < scene.patchList.size(); i++ ) {
            patchInit(scene.patchList.get(i));
        }

        galerkinState.topCluster = null;
        galerkinState.lastClusterId = -1;
        galerkinState.lastEye.set(Numeric.HUGE_FLOAT_VALUE, Numeric.HUGE_FLOAT_VALUE, Numeric.HUGE_FLOAT_VALUE);
    }

    @Override
    public boolean doStep(Scene scene, RenderOptions renderOptions) {
        if ( scene == null || renderOptions == null ) {
            // Parameters intentionally unused in this reduced Java migration.
        }

        if ( galerkinState.iterationNumber < 0 ) {
            Error.error("doGalerkinOneStep", "method not initialized");
            return true; // Done, don't continue!
        }

        galerkinState.iterationNumber++;

        return true;
    }

    /**
Disposes of the cluster hierarchy
*/
    private static void galerkinDestroyClusterHierarchy(GalerkinElement clusterElement) {
        if ( clusterElement == null || !clusterElement.isCluster() ) {
            return;
        }

        for ( int i = 0;
              clusterElement.irregularSubElements != null && i < clusterElement.irregularSubElements.size();
              i++ ) {
            Element child = clusterElement.irregularSubElements.get(i);
            if ( child instanceof GalerkinElement ) {
                galerkinDestroyClusterHierarchy((GalerkinElement)child);
            }
        }
    }

    @Override
    public void terminate(ArrayList<Patch> scenePatches) {
        if ( scenePatches != null ) {
            for ( int i = 0; i < scenePatches.size(); i++ ) {
                Patch patch = scenePatches.get(i);
                if ( patch != null ) {
                    recomputePatchColor(patch);
                }
            }
        }

        if ( galerkinState.topCluster != null ) {
            galerkinDestroyClusterHierarchy(galerkinState.topCluster);
            galerkinState.topCluster = null;
        }
        galerkinState.iterationNumber = -1;
    }

    @Override
    public ColorRgb getRadiance(
        Camera camera,
        Patch patch,
        double u,
        double v,
        Vector3D dir,
        RenderOptions renderOptions)
    {
        if ( camera != null || u != 0.0 || v != 0.0 || dir != null || renderOptions != null ) {
            // Parameters intentionally unused in this reduced Java migration.
        }

        if ( patch == null || !(patch.radianceData instanceof GalerkinElement) ) {
            ColorRgb black = new ColorRgb();
            black.clear();
            return black;
        }

        GalerkinElement element = (GalerkinElement)patch.radianceData;
        return element.radiance[0];
    }

    @Override
    public Element createPatchData(Patch patch) {
        GalerkinElement element = new GalerkinElement(patch, galerkinState);
        element.Ed = PatchVisitor.averageEmittance(patch, ALL_XXDF_COMPONENTS);
        element.Rd = PatchVisitor.averageNormalAlbedo(patch, ALL_BSDF_COMPONENTS);
        element.radiance[0].set(element.Ed.r, element.Ed.g, element.Ed.b);
        element.unShotRadiance[0].set(element.Ed.r, element.Ed.g, element.Ed.b);
        return element;
    }

    @Override
    public void destroyPatchData(Patch patch) {
        if ( patch != null ) {
            patch.radianceData = null;
        }
    }

    @Override
    public String getStats() {
        StringBuilder stats = new StringBuilder(STRING_LENGTH);
        stats.append("Galerkin\nStatistics\n\n");
        stats.append(String.format(Locale.US, "Iteration nr: %d\n", galerkinState.iterationNumber));
        stats.append(String.format(Locale.US, "CPU time: %g secs\n", galerkinState.cpuSeconds));
        stats.append(String.format(Locale.US, "Clustered: %d\n", galerkinState.clustered));
        stats.append(String.format(Locale.US, "Importance driven: %d\n", galerkinState.importanceDriven));
        return stats.toString();
    }

    @Override
    public void writeVRML(
        Camera camera,
        OutputStream outputStream,
        RenderOptions renderOptions)
    {
        if ( camera == null || outputStream == null || renderOptions == null ) {
            return;
        }

        VrmlWriter.writeHeader(camera, outputStream, renderOptions);
        VrmlWriter.writeTrailer(outputStream);
    }
}

File: turboc/base/src/galerkin/GalerkinRadianceMethod.cpp
/**
Galerkin radiosity, with the following variants:
- With or without hierarchical refinement
- With or without clusters
- With Jacobi, Gauss-Seidel or South well iterations
- With potential-driven or not
*/

#include <string.h>

#include "java/util/ArrayList.txx"
#include "java/util/Formatter.h"
#include "common/logging/Logger.h"
#include "common/statistics/Statistics.h"
#include "common/color/Cie.h"
#include "tonemap/ToneMap.h"
#include "io/wrapper/PersistenceElement.h"
#include "io/wrl/VrmlWriter.h"
#include "galerkin/GalerkinBasis.h"
#include "galerkin/GalerkinRadianceMethod.h"
#include "galerkin/processing/ClusterCreationStrategy.h"
#include "galerkin/processing/GatheringClusteredStrategy.h"
#include "galerkin/processing/GatheringSimpleStrategy.h"
#include "galerkin/processing/ScratchVisibilityStrategy.h"
#include "galerkin/processing/ShootingStrategy.h"

GalerkinState GalerkinRadianceMethod::galerkinState;
OutputStream *GalerkinRadianceMethod::vrmlOutputStream = NULL;
int GalerkinRadianceMethod::numberOfWrites = 0;
int GalerkinRadianceMethod::vertexId = 0;

String
GalerkinRadianceMethod::formatToString(const char *format, va_list arguments) {
    if ( format == NULL ) {
        return String();
    }

    char localBuffer[256];
    va_list argumentsCopy;
    va_copy(argumentsCopy, arguments);
    const int required = Formatter::vformat(localBuffer, ((int)(sizeof(localBuffer))), format, argumentsCopy);
    va_end(argumentsCopy);

    if ( required < 0 ) {
        return String();
    }
    if ( required < ((int)(sizeof(localBuffer))) ) {
        return String(localBuffer);
    }

    char *dynamicBuffer = new char[required + 1];
    va_copy(argumentsCopy, arguments);
    Formatter::vformat(dynamicBuffer, required + 1, format, argumentsCopy);
    va_end(argumentsCopy);

    String result(dynamicBuffer);
    delete[] dynamicBuffer;
    return result;
}

void
GalerkinRadianceMethod::writeFormatted(const char *format, ...) {
    if ( vrmlOutputStream == NULL || format == NULL ) {
        return;
    }

    va_list arguments;
    va_start(arguments, format);
    String text = GalerkinRadianceMethod::formatToString(format, arguments);
    va_end(arguments);

    if ( text.isEmpty() ) {
        return;
    }

    PersistenceElement::writeBytes(
        *vrmlOutputStream,
        ((const unsigned char *)(text.toCString())),
        text.length());
}

void
GalerkinRadianceMethod::appendStatsText(char *buffer, int *offset, const char *format, ...) {
    if ( *offset >= STRING_LENGTH - 1 ) {
        return;
    }

    va_list arguments;
    va_start(arguments, format);
    String text = GalerkinRadianceMethod::formatToString(format, arguments);
    va_end(arguments);

    const int written = text.length();
    if ( written <= 0 ) {
        return;
    }

    const int available = STRING_LENGTH - *offset - 1;
    const int copied = (written > available) ? available : written;
    if ( copied > 0 ) {
        memcpy(&buffer[*offset], text.toCString(), ((size_t)(copied)));
        *offset += copied;
        buffer[*offset] = '\0';
    }
}

void
GalerkinRadianceMethod::writeVertexCoord(const Vector3D *p) {
    if ( numberOfWrites > 0 ) {
        writeFormatted("%s", ", ");
    }
    numberOfWrites++;
    if ( numberOfWrites % 4 == 0 ) {
        writeFormatted("%s", "\n\t  ");
    }
    writeFormatted("%g %g %g", p->x, p->y, p->z);
    vertexId++;
}

void
GalerkinRadianceMethod::writeVertexCoords(Element *element) {
    const GalerkinElement *galerkinElement = ((GalerkinElement *)(element));
    Vector3D v[8];
    int numberOfVertices = galerkinElement->vertices(v);
    for ( int i = 0; i < numberOfVertices; i++ ) {
        writeVertexCoord(&v[i]);
    }
}

void
GalerkinRadianceMethod::writeCoords() {
    numberOfWrites = vertexId = 0;
    writeFormatted("%s", "\tcoord Coordinate {\n\t  point [ ");
    GalerkinRadianceMethod::galerkinState.topCluster->traverseAllLeafElements(writeVertexCoords);
    writeFormatted("%s", " ] ");
    writeFormatted("%s", "\n\t}\n");
}

void
GalerkinRadianceMethod::writeVertexColor(const ColorRgbMutable *color) {
    if ( numberOfWrites > 0 ) {
        writeFormatted("%s", ", ");
    }
    numberOfWrites++;
    if ( numberOfWrites % 4 == 0 ) {
        writeFormatted("%s", "\n\t  ");
    }
    writeFormatted("%.3g %.3g %.3g", color->getR(), color->getG(), color->getB());
    vertexId++;
}

void
GalerkinRadianceMethod::writeVertexColors(Element *element) {
    const GalerkinElement *galerkinElement = ((GalerkinElement *)(element));
    ColorRgbMutable vertexRadiosity[4];
    int i;

    if ( galerkinElement->patch->numberOfVertices == 3 ) {
        vertexRadiosity[0] = GalerkinBasis::radianceAtPoint(galerkinElement, galerkinElement->radiance, 0.0, 0.0);
        vertexRadiosity[1] = GalerkinBasis::radianceAtPoint(galerkinElement, galerkinElement->radiance, 1.0, 0.0);
        vertexRadiosity[2] = GalerkinBasis::radianceAtPoint(galerkinElement, galerkinElement->radiance, 0.0, 1.0);
    } else {
        vertexRadiosity[0] = GalerkinBasis::radianceAtPoint(galerkinElement, galerkinElement->radiance, 0.0, 0.0);
        vertexRadiosity[1] = GalerkinBasis::radianceAtPoint(galerkinElement, galerkinElement->radiance, 1.0, 0.0);
        vertexRadiosity[2] = GalerkinBasis::radianceAtPoint(galerkinElement, galerkinElement->radiance, 1.0, 1.0);
        vertexRadiosity[3] = GalerkinBasis::radianceAtPoint(galerkinElement, galerkinElement->radiance, 0.0, 1.0);
    }

    if ( GalerkinRadianceMethod::galerkinState.useAmbientRadiance ) {
        ColorRgbMutable reflectivity = galerkinElement->patch->radianceData->Rd;
        ColorRgbMutable ambient;

        ambient.scalarProduct(reflectivity, GalerkinRadianceMethod::galerkinState.ambientRadiance);
        for ( i = 0; i < galerkinElement->patch->numberOfVertices; i++ ) {
            vertexRadiosity[i].add(vertexRadiosity[i], ambient);
        }
    }

    for ( i = 0; i < galerkinElement->patch->numberOfVertices; i++ ) {
        ColorRgb col;
        ToneMap::radianceToRgb(ColorRgb(vertexRadiosity[i]), &col, *GalerkinRadianceMethod::galerkinState.toneMapOptions);
        ColorRgbMutable out(col.getR(), col.getG(), col.getB());
        writeVertexColor(&out);
    }
}

void
GalerkinRadianceMethod::writeVertexColorsTopCluster() {
    vertexId = numberOfWrites = 0;
    writeFormatted("%s", "\tcolor Color {\n\t  color [ ");
    GalerkinRadianceMethod::galerkinState.topCluster->traverseAllLeafElements(writeVertexColors);
    writeFormatted("%s", " ] ");
    writeFormatted("%s", "\n\t}\n");
}

void
GalerkinRadianceMethod::writeColors(const RenderOptions *renderOptions) {
    if ( !renderOptions->isSmoothShading() ) {
        Logger::warning(NULL, "I assume you want a smooth shaded model ...");
    }
    writeFormatted("\tcolorPerVertex %s\n", "TRUE");
    writeVertexColorsTopCluster();
}

void
GalerkinRadianceMethod::writeCoordIndex(int index) {
    numberOfWrites++;
    if ( numberOfWrites % 20 == 0 ) {
        writeFormatted("%s", "\n\t  ");
    }
    writeFormatted("%d ", index);
}

void
GalerkinRadianceMethod::writeCoordIndices(Element *element) {
    const GalerkinElement *galerkinElement = ((GalerkinElement *)(element));
    for ( int i = 0; i < galerkinElement->patch->numberOfVertices; i++ ) {
        writeCoordIndex(vertexId);
        vertexId++;
    }
    writeCoordIndex(-1);
}

void
GalerkinRadianceMethod::writeCoordIndicesTopCluster() {
    vertexId = numberOfWrites = 0;
    writeFormatted("%s", "\tcoordIndex [ ");
    GalerkinRadianceMethod::galerkinState.topCluster->traverseAllLeafElements(writeCoordIndices);
    writeFormatted("%s", " ]\n");
}

void
GalerkinRadianceMethod::freeMemory() {
    if ( GalerkinRadianceMethod::galerkinState.scratch != NULL ) {
        delete GalerkinRadianceMethod::galerkinState.scratch;
        GalerkinRadianceMethod::galerkinState.scratch = NULL;
    }
}

GalerkinRadianceMethod::GalerkinRadianceMethod() {
    className = GALERKIN;
    gatheringStrategy = NULL;
}

GalerkinRadianceMethod::~GalerkinRadianceMethod() {
    if ( galerkinState.topCluster != NULL ) {
        galerkinState.topCluster = NULL;
    }
    if ( gatheringStrategy != NULL ) {
        delete gatheringStrategy;
        gatheringStrategy = NULL;
    }
}

/**
For counting how much CPU time was used for the computations
*/
void
GalerkinRadianceMethod::updateCpuSecs() {
    const long t = System::nanoTime();
    galerkinState.cpuSeconds += ((float)(((double)(t - galerkinState.lastClock)) / 1000000000.0));
    galerkinState.lastClock = t;
}

void
GalerkinRadianceMethod::patchInit(Patch *patch) {
    ColorRgbMutable reflectivity = patch->radianceData->Rd;
    ColorRgbMutable selfEmittanceRadiance = patch->radianceData->Ed;

    if ( galerkinState.useConstantRadiance ) {
        // See Neumann et-al, "The Constant Radiosity Step", Euro-graphics Rendering Workshop
        // '95, Dublin, Ireland, June 1995, p 336-344
        galerkinGetRadiance(patch).scalarProduct(reflectivity, galerkinState.constantRadiance);
        galerkinGetRadiance(patch).add(galerkinGetRadiance(patch), selfEmittanceRadiance);
        if ( galerkinState.galerkinIterationMethod == SOUTH_WELL ) {
            patch->radianceData->unShotRadiance[0].subtract(galerkinGetRadiance(patch), galerkinState.constantRadiance);
        }
    } else {
        galerkinSetRadiance(patch, selfEmittanceRadiance);
        if ( galerkinState.galerkinIterationMethod == SOUTH_WELL ) {
            patch->radianceData->unShotRadiance[0] = galerkinGetRadiance(patch);
        }
    }

    if ( galerkinState.importanceDriven ) {
        switch ( galerkinState.galerkinIterationMethod ) {
            case GAUSS_SEIDEL:
            case JACOBI:
                galerkinSetPotential(patch, patch->directPotential);
                break;
            case SOUTH_WELL:
                galerkinSetPotential(patch, patch->directPotential);
                galerkinSetUnShotPotential(patch, patch->directPotential);
                break;
            default:
                Logger::fatal(-1, "patchInit", "Invalid iteration method");
        }
    }

    recomputePatchColor(patch);
}

/**
Recomputes the color of a patch using ambient radiance term, ... if requested for
*/
void
GalerkinRadianceMethod::recomputePatchColor(Patch *patch) {
    ColorRgbMutable reflectivity = patch->radianceData->Rd;
    ColorRgbMutable radVis;

    // Compute the patches color based on its radiance + ambient radiance if desired
    if ( galerkinState.useAmbientRadiance ) {
        radVis.scalarProduct(reflectivity, galerkinState.ambientRadiance);
        radVis.add(radVis, galerkinGetRadiance(patch));
        ToneMap::radianceToRgb(ColorRgb(radVis), &patch->color, *galerkinState.toneMapOptions);
    } else {
        ToneMap::radianceToRgb(ColorRgb(galerkinGetRadiance(patch)), &patch->color, *galerkinState.toneMapOptions);
    }
    patch->computeVertexColors();
}

const char *
GalerkinRadianceMethod::getRadianceMethodName() const  {
    return "Galerkin";
}

void
GalerkinRadianceMethod::setStrategy() {
    if ( galerkinState.clustered ) {
        gatheringStrategy = new GatheringClusteredStrategy();
    } else {
        gatheringStrategy = new GatheringSimpleStrategy();
    }
}

void
GalerkinRadianceMethod::parseOptions(int * /*argc*/, char ** /*argv*/) {
}

void
GalerkinRadianceMethod::initialize(Scene *scene, ToneMappingContext *toneMapOptions) {
    galerkinState.toneMapOptions = toneMapOptions;
    if ( galerkinState.toneMapOptions == NULL ) {
        Logger::fatal(-1, "GalerkinRadianceMethod::initialize", "Tone mapping context not provided");
    }

    galerkinState.iterationNumber = 0;
    galerkinState.cpuSeconds = 0.0;

    GalerkinElement::initializeBasis();

    galerkinState.constantRadiance = ColorRgbMutable(Statistics::instance().radiance.estimatedAverageRadiance.getR(), Statistics::instance().radiance.estimatedAverageRadiance.getG(), Statistics::instance().radiance.estimatedAverageRadiance.getB());
    if ( galerkinState.useConstantRadiance ) {
        galerkinState.ambientRadiance = ColorRgbMutable(0.0f, 0.0f, 0.0f);
    } else {
        galerkinState.ambientRadiance = ColorRgbMutable(Statistics::instance().radiance.estimatedAverageRadiance.getR(), Statistics::instance().radiance.estimatedAverageRadiance.getG(), Statistics::instance().radiance.estimatedAverageRadiance.getB());
    }

    for ( int i = 0; scene->patchList != NULL && i < scene->patchList->size(); i++ ) {
        patchInit(scene->patchList->get(i));
    }

    galerkinState.topCluster = ClusterCreationStrategy::createClusterHierarchy(
        scene->clusteredRootGeometry, &galerkinState);

    // Create a scratch software renderer for various operations on clusters
    if ( galerkinState.clusteringStrategy == Z_VISIBILITY ) {
        ScratchVisibilityStrategy::scratchInit(&galerkinState);
    }

    // Global variables for scratch rendering
    galerkinState.lastClusterId = -1;
    galerkinState.lastEye.set(Numeric::HUGE_FLOAT_VALUE, Numeric::HUGE_FLOAT_VALUE, Numeric::HUGE_FLOAT_VALUE);
}

bool
GalerkinRadianceMethod::doStep(Scene *scene, RenderOptions *renderOptions) {
    if ( galerkinState.iterationNumber < 0 ) {
        Logger::error("doGalerkinOneStep", "method not initialized");
        return true; // Done, don't continue!
    }

    galerkinState.iterationNumber++;
    galerkinState.lastClock = System::nanoTime();

    // And now the real work
    int done = 1;

    switch ( galerkinState.galerkinIterationMethod ) {
        case JACOBI:
        case GAUSS_SEIDEL:
            done = gatheringStrategy->doGatheringIteration(scene, &galerkinState, renderOptions);
            break;
        case SOUTH_WELL:
            done = ShootingStrategy::doShootingStep(scene, &galerkinState, renderOptions);
            break;
        default:
            Logger::fatal(2, "doGalerkinOneStep", "Invalid iteration method %d\n", galerkinState.galerkinIterationMethod);
    }

    updateCpuSecs();

    return done;
}

/**
Disposes of the cluster hierarchy
*/
void
GalerkinRadianceMethod::galerkinDestroyClusterHierarchy(GalerkinElement *clusterElement) {
    if ( !clusterElement || !clusterElement->isCluster() ) {
        return;
    }

    for ( int i = 0; clusterElement->irregularSubElements != NULL && i < clusterElement->irregularSubElements->size(); i++ ) {
        GalerkinRadianceMethod::galerkinDestroyClusterHierarchy(((GalerkinElement *)(clusterElement->irregularSubElements->get(i))));
    }
    delete clusterElement;
}

void
GalerkinRadianceMethod::terminate(ArrayList<Patch *> */*scenePatches*/) {
    if ( galerkinState.clusteringStrategy == Z_VISIBILITY ) {
        ScratchVisibilityStrategy::scratchTerminate(&galerkinState);
    }
    if ( galerkinState.topCluster != NULL ) {
        GalerkinRadianceMethod::galerkinDestroyClusterHierarchy(galerkinState.topCluster);
        delete galerkinState.topCluster;
        galerkinState.topCluster = NULL;
    }
}

ColorRgbMutable
GalerkinRadianceMethod::computePatchRadiance(Patch *patch, double u, double v) const
{
    const GalerkinElement *leaf;
    ColorRgbMutable rad;

    if ( patch->jacobian ) {
        patch->biLinearToUniform(&u, &v);
    }

    GalerkinElement *topLevelElement = GalerkinElement::fromPatch(patch);
    leaf = topLevelElement->regularLeafAtPoint(&u, &v);

    rad = GalerkinBasis::radianceAtPoint(leaf, leaf->radiance, u, v);

    if ( galerkinState.useAmbientRadiance ) {
        // Add ambient radiance
        ColorRgbMutable reflectivity = patch->radianceData->Rd;
        ColorRgbMutable ambientRadiance;
        ambientRadiance.scalarProduct(reflectivity, galerkinState.ambientRadiance);
        rad.add(rad, ambientRadiance);
    }

    return rad;
}

ColorRgbMutable
GalerkinRadianceMethod::getRadiance(
    Camera * /*camera*/,
    Patch *patch,
    double u,
    double v,
    Vector3D /*dir*/,
    const RenderOptions * /*renderOptions*/) const
{
    return computePatchRadiance(patch, u, v);
}

/**
Radiance data for a Patch is a surface element
*/
Element *
GalerkinRadianceMethod::createPatchData(Patch *patch) {
    return patch->radianceData = new GalerkinElement(patch, &galerkinState);
}

void
GalerkinRadianceMethod::destroyPatchData(Patch *patch) {
    if ( patch == NULL ) {
        return;
    }
    if ( patch->radianceData != NULL ) {
        delete ((GalerkinElement *)(patch->radianceData));
        patch->radianceData = NULL;
    }
}

char *
GalerkinRadianceMethod::getStats() const {
    static char stats[STRING_LENGTH] = {0};

    for ( int i = 0 ; i < STRING_LENGTH; i++ ) {
        stats[i] = '\0';
    }

    int statsOffset = 0;

    appendStatsText(stats, &statsOffset, "Galerkin Radiosity Statistics:\n\n");
    appendStatsText(stats, &statsOffset, "Iteration: %d\n\n", galerkinState.iterationNumber);
    appendStatsText(stats, &statsOffset, "Nr. elements: %d\n", GalerkinElement::getNumberOfElements());
    appendStatsText(stats, &statsOffset, "clusters: %d\n", GalerkinElement::getNumberOfClusters());
    appendStatsText(stats, &statsOffset, "surface elements: %d\n\n", GalerkinElement::getNumberOfSurfaceElements());
    appendStatsText(stats, &statsOffset, "Nr. interactions: %d\n", Interaction::getNumberOfInteractions());
    appendStatsText(stats, &statsOffset, "cluster to cluster: %d\n", Interaction::getNumOClustTClustInters());
    appendStatsText(stats, &statsOffset, "cluster to surface: %d\n", Interaction::getNumOClustTSurfInters());
    appendStatsText(stats, &statsOffset, "surface to cluster: %d\n", Interaction::getNumOSurfTClustInters());
    appendStatsText(stats, &statsOffset, "surface to surface: %d\n", Interaction::getNumOSurfTSurfInters());
    appendStatsText(stats, &statsOffset, "shadow hits: %d\n", Statistics::instance().shadow.numberOfShadowRays);
    appendStatsText(stats, &statsOffset, "shadow hits cached: %d\n", Statistics::instance().shadow.numberOfShadowCacheHits);
    appendStatsText(stats, &statsOffset, "CPU time: %g secs.\n", galerkinState.cpuSeconds);
    appendStatsText(stats, &statsOffset, "Minimum element area: %g m^2\n", Statistics::instance().radiance.totalArea * ((double)(galerkinState.relMinElemArea)));
    appendStatsText(stats, &statsOffset, "Link error threshold: %g %s\n\n",
         (galerkinState.errorNorm == RADIANCE_ERROR ?
                   M_PI * (galerkinState.relLinkErrorThreshold *
                           Cie::spectrumLuminance(Statistics::instance().radiance.maxSelfEmittedRadiance.getR(), Statistics::instance().radiance.maxSelfEmittedRadiance.getG(), Statistics::instance().radiance.maxSelfEmittedRadiance.getB())) :
                   galerkinState.relLinkErrorThreshold *
                   Cie::spectrumLuminance(Statistics::instance().radiance.maxSelfEmittedPower.getR(), Statistics::instance().radiance.maxSelfEmittedPower.getG(), Statistics::instance().radiance.maxSelfEmittedPower.getB())),
         (galerkinState.errorNorm == RADIANCE_ERROR ? "lux" : "lumen"));

    return stats;
}

void
GalerkinRadianceMethod::writeVRML(
    const Camera *camera,
    OutputStream *outputStream,
    const RenderOptions *renderOptions) const
{
    if ( outputStream == NULL ) {
        return;
    }
    VrmlWriter::writeHeader(camera, outputStream, renderOptions);

    vrmlOutputStream = outputStream;
    writeCoords();
    writeColors(renderOptions);
    writeCoordIndicesTopCluster();

    VrmlWriter::writeTrailer(outputStream);
}
